### ISSUE TITLE
react-router-v4?

### ISSUE THREAD

**Issue by ronag** [reactjs/react-router-redux#454]

**Comment by timdorr** [comment]
No need. Because everything is now declarative (see `react-history`), we can use render functions to sync the router state into redux. This library won't be needed for 4.0.

**Comment by mxstbr** [comment]
Is it possible to dispatch `LOCATION_CHANGE` actions when a location is changed? Together with the `redux-saga` watcher pattern this has enabled me to take much better care of my long running transactions, and I'd hate for this use case to be overlooked.

**Comment by timdorr** [comment]
I don't know what happened to it, but there was a Redux example for v4: https://github.com/ReactTraining/react-router/blob/7f6706dab4827afc1c26a58418f8ef8c8ab40125/website/examples/Redux.js

**Comment by pedrofelixsky** [comment]
That example seems to be using the `BrowserRouter` with the `location` provided by the redux state and the `onChange` dispatching an event to redux. However from the latest v4 sources, `BrowserRouter` does not have those properties. Shouldn't `BrowserRouter` be replaced by `StaticRouter` on the example? Am I missing something?

**Comment by choonchernlim** [comment]
Based on that example, I can't get mine sample app to work either due to what @pedrofelixsky said. If I switch `BrowserRouter` to `StaticRouter`, then I'm getting error saying that `action`, `onPush` and `onReplace` are not provided to it. 

Would it be possible for someone to revise that Redux example for v4?

Thank you.

**Comment by timdorr** [comment]
I'll be making a 5.0 branch here soon and that will have an integration implementation and example.

**Comment by mkozhukharenko** [comment]
>I don't know what happened to it, but there was a Redux example for v4: https://github.com/ReactTraining/react-router/blob/7f6706dab4827afc1c26a58418f8ef8c8ab40125/website/examples/Redux.js

@timdorr but according to current source code there not `onChange` method in the `BrowserRouter`. Am I wrong?

**Comment by tomitrescak** [comment]
@timdorr 5.0 branch? There will be new version of react-router as well?

**Comment by timdorr** [comment]
https://github.com/ReactTraining/react-router/tree/v4

**Comment by tomitrescak** [comment]
I'm probably not following but that's v4?. And the example there does not work with redux. Sorrrryy to  be pain in ... :/

**Comment by jamiehill** [comment]
Also scratching head :/

**Comment by timdorr** [comment]
The version of react-router-redux currently out is 4.0. It's ahead of react router, which has 2.x out currently, 3.0 coming very very soon, and 4.0 following right behind that, which is a complete rewrite of the library. The next major version here is 5.0, which is needed to rewrite the library yet again to work with react router 4.0.

**Comment by jamiehill** [comment]
@timdorr ok, so what you have [here](https://github.com/timdorr/react-router-redux/tree/5.0.x) is the beginnings of the 5.0 implementation?  Unfortunately isn't pulling down all the files when `npm` installing atm, and I'm assuming the api is not concrete yet anyway?

**Comment by timdorr** [comment]
It's still a pull request (#460). I haven't finalized anything about it yet.

**Comment by jamiehill** [comment]
I think the confusion in my case was more based on the fact that the Redux example above, suggests that with react-router 4, react-router-redux would no longer be necessary.  But coupled with that, the api it uses in the example, doesn't seem to marry with the version it uses

**Comment by jamiehill** [comment]
Sorry that PR which I hadn't seen before, seems to shed some light on this

**Comment by timdorr** [comment]
The internal mechanics of React Router 4.0 have changed quite a bit since it was put out as a prerelease a couple months ago. So, the ability to more easily integrate something like Redux have gone away. It's still possible, but just be done another way. And we don't have to rely on undocumented or hacky APIs to do it, so it will be better this way.

**Comment by dolfelt** [comment]
For anyone still interested, I took the 5.x branch and modified it to work with the newer code pattern of react-router@next

It's simply two files that allows redux to track the location change. This doesn't allow you to use the redux router actions, but imo, with v4, the transitioning and routing should be a concern of the components.

https://gist.github.com/dolfelt/304e4a99b6f413d609994732a33b5906

**Comment by Murkrage** [comment]
@dolfelt I'm not really sure how to implement it.

**Comment by ryanflorence** [comment]
I haven't read the whole thread here (sorry! for real tl;dr!) but should probably give my thoughts.

I think this lib will still be around when router v4 ships, it'll be pretty tiny though. It will depend on https://github.com/ReactTraining/react-router-addons-controlled and re-export those routers along with a reducer and action for navigating by dispatching.

```js
import { BrowserRouter, navigate, locationReducer } from 'react-router-redux'
```

**Comment by lourd** [comment]
I've been working on this a couple days and wanted to share which solution I ended up going with.

**tl;dr** I used `StaticRouter` from `react-router` v4, created my own history from `history`, with minor changes to the current version of `react-redux-router`. I like this setup because URL state flows cleanly from history -> store -> router -> components.

Aside: can we re-open this issue? It's currently difficult to tell where the react-router v4 discussion is happening, it would help with visibility.

-------

This library _almost_ works out of the box with `StaticRouter`. The only thing that it doesn't handle appropriately is setting the initial state. That responsibility used to lie with the Router, now it lies within the reducer, which I believe makes more sense. The reducer must gets its initial state from the `history`, which is what the change was.

This is the old version
```js
// react-router-redux reducer.js
const LOCATION_CHANGE = '@@router/LOCATION_CHANGE'
const initialState = {
  locationBeforeTransitions: null,
}
export function routerReducer(state = initialState, { type, payload } = {}) {
  if (type === LOCATION_CHANGE) {
    return { ...state, locationBeforeTransitions: payload }
  }

  return state
}
```

This is the new version. It exports a factory function that takes a `history` object in order to initialize the state.
```js
const LOCATION_CHANGE = '@@router/LOCATION_CHANGE'
export const createReducer = history => {
  const initState = {
    location: history.location,
    action: history.action,
  }
  return function reducer(state = initState, action) {
    if (action.type === LOCATION_CHANGE) {
      return { ...state, location: action.payload }
    }

    return state
  }
}
```

Then in my root reducer file:
```js
// reducers/index.js
import { combineReducers } from 'redux'
// Using the new factory function
import { createReducer } from 'react-router-redux'
import createBrowserHistory from 'history/createBrowserHistory'

export const history = createBrowserHistory()
const routerReducer = createReducer(history)
export default combineReducers({
  ...reducers,
  routing: routerReducer,
})
```

The store setup works the same as it does currently, with `syncHistoryWithStore`
```js
// store.js
import { createStore, applyMiddleware, compose } from 'redux'
import {
  syncHistoryWithStore,
  routerMiddleware as createRouterMiddleware,
} from 'react-router-redux'

import rootReducer, { history } from './reducers/'
// Same API as before
const routerMiddleware = createRouterMiddleware(history)

// Enhance createStore with middlewares
const enhancedCreateStore = compose(
  applyMiddleware(routerMiddleware),
  window.__REDUX_DEVTOOLS_EXTENSION__
    ? window.__REDUX_DEVTOOLS_EXTENSION__()
    : f => f,
)(createStore)

const store = enhancedCreateStore(rootReducer)
// Same API
syncHistoryWithStore(history, store)

export default store
```

Hooking up the App needs some small wiring using Redux's `connect` decorator.
```js
// App.js
import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import { Match, Link, StaticRouter as Router } from 'react-router'
import { push, replace } from 'react-router-redux'

function mapStateToProps(state) {
  return {
    location: state.routing.location,
    action: state.routing.action,
  }
}

const mapDispatchToProps = {
  onPush: push,
  onReplace: replace,
}

@connect(mapStateToProps, mapDispatchToProps)
class App extends Component {
  static propTypes = {
    location: PropTypes.object.isRequired,
    action: PropTypes.string.isRequired,
    onPush: PropTypes.func.isRequired,
    onReplace: PropTypes.func.isRequired,
  }

  render() {
    return (
      <Router
        location={this.props.location}
        action={this.props.action}
        onPush={this.props.onPush}
        onReplace={this.props.onReplace}
      >
        <div>
          <div><Link to="/">Home</Link></div>
          <div><Link to="/foo">Foo</Link></div>
          <div><Link to="/bar">Bar</Link></div>
          <div>
            <Match pattern="/" exactly component={render={() => <h1>Home!</h1>}} />
            <Match pattern="/foo" exactly render={() => <h1>Foo!</h1>} />
            <Match pattern="/bar" exactly render={() => <h1>Bar!</h1>} />
          </div>
        </div>
      </Router>
    )
  }
}

export default App
```

This works better than the `ControlledRouter` from [react-router-addons-controlled](https://github.com/ReactTraining/react-router-addons-controlled). That had some bugs with time travel debugging (see https://www.youtube.com/watch?v=45WpW7E5CyQ). When I dug into the source of `ControlledRouter` I realized it was duplicating a lot of the logic that `sync` does from this library, which is what brought me to trying this solution.

I haven't tried all of the options (namely setting a basename or using hash history), but so far so good.

**Comment by ryanflorence** [comment]
Looks great, I'd love to ditch ControlledRouter since the only point of it really is to make redux users happy (and it's kind of fun to try to make work).

I know that this lib screws up the dropdown menus in "forward" and "back".

Also, can you navigate by dispatching or does this simply keep your store in sync w/ the url?

**Comment by lourd** [comment]
Technically everything is navigating by dispatching in this setup.

1. Link click invokes callbacks from the Router
2. Router calls its respective `onPush` or `onReplace` prop
3. Those props dispatch `CALL_HISTORY_METHOD` actions which get caught by the middleware that calls the corresponding `history` method
4. The special synced history dispatches `@@router/LOCATION_CHANGE` actions which the reducer sees and updates its state, which flows back down to the components.

The one thing I'm probably handling wrong is the `action` prop. Right now it's set an initialization and doesn't change again because the `@@router/LOCATION_CHANGE` action doesn't pass it along.

I also haven't thought through the "blocking navigation" API, but I'm pretty sure could also be handled with store state.

I think the browser forward & back dropdown menus work normally? From what I can tell. Not sure what I'd be looking for.

**Comment by ryanflorence** [comment]
> I think the browser forward & back dropdown menus work normally? From what I can tell. Not sure what I'd be looking for.

You can click + hold, then go back 3, 5, 10 entries. Same with forward. This lib just uses `replace` rather than `go(delta)`, so when you jump back 5 entries you blow away the 4 you jumped over, so the forward menu is empty.

**Comment by lourd** [comment]
Interesting... I'm not seeing that. I guess the different flow fixes it? ¯\_(ツ)_/¯ The back and forward menus invoke the respective `history` method, which dispatches a generic `@@router/LOCATION_CHANGE` action, changing the `location` in the store.

**Comment by ryanflorence** [comment]
eh, I'm sure if you had a demo I could break it :P but again, not a big deal I don't think.

**Comment by timdorr** [comment]
Hey, I'm leaving this closed to track work in #460

**Comment by collegepinger** [comment]
Is there Any possibility of doing SSR with Redux React-router 4
